File: 64. Minimum Path Sum/main.py
from typing import List


class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        rowSize = len(grid)
        colSize = len(grid[0])

        for row in range(1, rowSize):
            grid[row][0] += grid[row - 1][0]

        for col in range(1, colSize):
            grid[0][col] += grid[0][col - 1]

        for row in range(1, rowSize):
            for col in range(1, colSize):
                grid[row][col] += min(grid[row - 1][col], grid[row][col - 1])

        return grid[rowSize - 1][colSize - 1]

File: 64. Minimum Path Sum/main.rs
struct Solution;

impl Solution {
    pub fn min_path_sum(grid: Vec<Vec<i32>>) -> i32 {
        let row_size = grid.len();
        let col_size = grid[0].len();
        let mut grid = grid;

        for row in 1..row_size {
            grid[row][0] += grid[row - 1][0];
        }

        for col in 1..col_size {
            grid[0][col] += grid[0][col - 1];
        }

        for row in 1..row_size {
            for col in 1..col_size {
                grid[row][col] += grid[row - 1][col].min(grid[row][col - 1]);
            }
        }

        grid[row_size - 1][col_size - 1]
    }
}
